### PR TITLE
Login modification error handling

### PR DESCRIPTION
**Pull request by bgarciaentornos**
Also added for free some login unit testing, as there wasn't any (and didn't remember errors were being handled on a different class).

### CODE REVIEW

**Review by buckett — COMMENTED** [review]
Is it possible to make this error handling more generic as the errors object that is returned from most endpoints is pretty much the same. I raised this on https://github.com/oxctl/canvas-api/pull/29#issuecomment-777596491 and ideally I think we should address it now.

**Comment by bgarciaentornos** [comment]
I understand what you mean, but from what I've seen the formats for user and login errors are similar but still different enough (different levels of json branching) to reflect them on static classes. Do you know how to do this in a dynamic way that will still allow retrieving the most outstanding fields from the reponse for each different json format?

**Comment by buckett** [comment]
Can't you have a `Map<String,List<ErrorDetails>>` field. This way any error object should be able to map to it.

**Comment by bgarciaentornos** [comment]
```
  "errors": {
    "user": {
      "pseudonym": {
        "unique_id": <THIS IS WHAT WE COULD CALL ERRORDETAILS>

```
vs

```
`{
	"errors": {
		"unique_id": [{` <THIS IS WHAT WE COULD CALL ERRORDETAILS>

```

So the end ErrorDetails class might be the same, but they're on a different json level for this 2 handlers. Not sure for any other case though.

**Comment by bgarciaentornos** [comment]
Some updates:

- This new handler is generic, as it seems it will be valid for many endpoints. Valid patterns for now are those of login and course creation/update.
- Added ErrorDetails class with common attributes which can be used both by the generic and the user error handlers.
- Ideally the library should see if any 400 error can be transformed into an errors object (independently of endpoint pattern), and then throw the errors object wrapped in an exception.

File: src/main/java/edu/ksu/canvas/errors/GenericErrorHandler.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+import java.io.IOException;
+import java.util.regex.Pattern;
+
+/**
+ * The error handler that should be used when creation of a Canvas object fails.
+ */
+public class GenericErrorHandler implements ErrorHandler {
+
+    private Pattern coursePattern = Pattern.compile("/api/v1/accounts/\\d+/courses");
+    private Pattern loginPattern = Pattern.compile("/api/v1/accounts/\\d+/logins");
```
Comment (buckett) [inline src/main/java/edu/ksu/canvas/errors/GenericErrorHandler.java:19]: The URLs can have SIS IDs in them like `sis_course_id:an-example-sis-id` which wouldn't match on just the digits.

Comment (buckett) [reply src/main/java/edu/ksu/canvas/errors/GenericErrorHandler.java:19]: https://canvas.instructure.com/doc/api/file.object_ids.html

Comment (bgarciaentornos) [reply src/main/java/edu/ksu/canvas/errors/GenericErrorHandler.java:19]: Ok, I thought we weren't really covering all the cases on this PR, just adding a couple more that I had already tested. You proposed to move from this pattern check to every 400 error in the future, should I take a look to it now?

Comment (bgarciaentornos) [reply src/main/java/edu/ksu/canvas/errors/GenericErrorHandler.java:19]: I'll change it for "/accounts/xx/courses" but "/courses/yy/whatever" won't be covered as it is, until we add more patterns or catch the 400s.

**Pull request merged**